File: 1002-find-common-characters/1002-find-common-characters.py
class Solution:
  def commonChars(self, words: List[str]) -> List[str]:
    base = None
    for word in words:
      curr = defaultdict(int)
      for chr in word:
        curr[chr] += 1
      if base:
        for chr in base:
          if chr not in curr:
            base[chr] = 0
          elif base[chr] > curr[chr]:
            base[chr] = curr[chr]
      else:
        base = curr
    res = []
    for chr in base:
      for _ in range(base[chr]):
        res.append(chr)
    return res



        

File: 1002-find-common-characters/1002-find-common-characters.js
const commonChars = (words) => {
  let d = {}
  for (const letter of words[0]) {
    d[letter] = (d[letter] ?? 0) + 1
  }

  for (let i = 1; i < words.length; i++) {
    _d = {}
    for (const letter of words[i]) {
      if (d[letter]) {
        _d[letter] = (_d[letter] ?? 0) + 1
        d[letter]--
      }
    }
    if (Object.keys(_d).length === 0) return []
    d = _d
  }

  const res = []
  for (const letter in d) {
    while(d[letter]) {
      res.push(letter)
      d[letter]--
    }
  }
  return res
};